### PR TITLE
Ports - Kim

### PR DESCRIPTION
**Pull request by Kimberly-Fasbender**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | A lot of the basic functionality from Ruby is there in JavaScript (loops, conditional statements, data structures, etc), so I was able to use all of the foundational knowledge and apply it.
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Yes, and no. I was able to utilize Stack Overflow for both Ruby and Javascript, and Google searching. The only real difference for finding information online for me was using the MDN as a resource instead of Ruby Docs.
What was a challenge you were able to overcome on this assignment? | Finding out that JavaScript vanilla does not have a built-in method for pulling multiple random samples from an array (like Ruby does). I was able to overcome that challenge by utilizing a JavaScript library (lodash). 
What has been interesting and positive about learning a new programming language? | It's like learning a new spoken language in a way. It opens up new doors that weren't open before. I'm doing the same project as before, but it feels completely different and that's interesting and challenging in a good way. I'm excited to see more of what JavaScript can do in the browser. 👍 
What is something to focus on your learnings in JavaScript in the next week? | Learning more about the built-in methods that are available for use in JavaScript, getting more comfortable iterating over an object, and learning more about how to re-format from one data structure to another (taking an object and turning it into an array, etc).

### CODE REVIEW

**Comment by tildeee** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | hahahaha :+1:
**Code Requirements** |
`drawLetters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `drawLetters` pass | x
`usesAvailableLetters` method | x
All tests for `usesAvailableLetters` pass | x
`scoreWord` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `scoreWord` pass | x
`highestScoreFrom` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highestScoreFrom` pass | x
**Overall** |



Well done on this project, Kim! Your code looks pretty great! It's clean and the solutions are exactly what I was looking for.


Actually, the biggest comments I have on your code is that your variables tend to not be named very clearly, hahaha. I've also left a couple of comments to help your code get more JavaScript-y.


Otherwise, in general, your project looks good. Well done!

File: src/adagrams.js [hunk 1]
```diff
@@ -1,7 +1,104 @@
...
+    3: ['B', 'C', 'M', 'P'],
+    4: ['F', 'H', 'V', 'W', 'Y'],
+    5: ['K'],
+    8: ['J', 'X'],
+    10: ['Q', 'Z']
+  },
+
   drawLetters() {
-    // Implement this method for wave 1
+    return _.sampleSize(this.letters, 10);
```
Comment (tildeee) [inline src/adagrams.js:41]: NICE! I love a one-line method :) :) :)

File: src/adagrams.js [hunk 2]
```diff
@@ -1,7 +1,104 @@
...
+  },
+
+  usesAvailableLetters(input, lettersInHand) {
+    const playedLetters = (input.toUpperCase()).split('');
```
Comment (tildeee) [inline src/adagrams.js:45]: Nice! The parens definitely help with readability, but in case you wanted this to be a slimmer letter, you can rely on the fact that the code gets executed left-to-right, so this is the same code: `const playedLetters = input.toUpperCase().split('');`

File: src/adagrams.js [hunk 3]
```diff
@@ -1,7 +1,104 @@
...
+    let valid = {};
```
Comment (tildeee) [inline src/adagrams.js:46]: Strangely enough, because you never re-assign `valid` (you just modify its members), `valid` can be a `const`

Comment (tildeee) [inline src/adagrams.js:46]: `valid` is not my favorite variable name... Is there some other more accurate name you could use?

File: src/adagrams.js [hunk 4]
```diff
@@ -1,7 +1,104 @@
...
+    let result = true;
+
+    lettersInHand.forEach((letter) => {
+      valid[letter] ? valid[letter] += 1 : valid[letter] = 1;
```
Comment (tildeee) [inline src/adagrams.js:50]: NICE TERNARY!

File: src/adagrams.js [hunk 5]
```diff
@@ -1,7 +1,104 @@
...
+      } else {
+        result = false;
+      }
+    });
+    return result;
   },
+
+  scoreWord(word) {
+    const input = (word.toUpperCase()).split('');
+    const array = Object.entries(this.scores);
```
Comment (tildeee) [inline src/adagrams.js:65]: Hm, is there a better name for this? You could still use `scores` (as it is not the same as `this.scores`)

File: src/adagrams.js [hunk 6]
```diff
@@ -1,7 +1,104 @@
...
+    let total = 0;
+
+    input.forEach((letter) => {
+      array.forEach((score) => {
+        if (score[1].includes(letter)) {
```
Comment (tildeee) [inline src/adagrams.js:70]: This ends up reading weirdly (the `score[1]` bit)... There's no obvious way to make this read weirdly without refactoring the whole solution, but consider restructuring `this.scores` so that you won't have to rely on the structure that `Object.entries` gives you

File: src/adagrams.js [hunk 7]
```diff
@@ -1,7 +1,104 @@
...
+
+    if (word.length >= 7) {
+      total += 8;
+    }
+
+    return total;
+  },
+
+  highestScoreFrom(words) {
+    let highestScore = { word: '', score: 0};
```
Comment (tildeee) [inline src/adagrams.js:84]: as mentioned above, because `highestScore` is never re-assigned (only modified), it can be `const`
